flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/CastRuleUtils.java

```diff
@@ -296,6 +318,35 @@ public CodeWriter appendBlock(String codeBlock) {
             return this;
         }
 
+        public CodeWriter appendAndTrimStringIfNeeded(
+                boolean legacyBehaviour,
+                int precision,
+                String resultStringTerm,
+                String builderTerm) {
+            CodeWriter writer =
+                    declStmt(String.class, resultStringTerm)
+                            .assignStmt(resultStringTerm, methodCall(builderTerm, "toString"));
+
+            // Trim if needed
+            if (!legacyBehaviour && shouldPossiblyTrim(precision)) {
+                writer.ifStmt(
+                        lengthExceedsPrecision(builderTerm, precision),
+                        thenWriter ->
+                                thenWriter.assignStmt(
+                                        resultStringTerm,
+                                        methodCall(
+                                                builderTerm,
+                                                "substring",
+                                                0,
+                                                staticCall(
+                                                        Math.class,
+                                                        "min",
+                                                        methodCall(builderTerm, "length"),
+                                                        precision))));
+            }
+            return writer;
+        }
```

Comment:
I don't like much this method, it's doing way too many things, in particular in this context or `CodeWriter` instance methods where my goal is to just have basic java constructs, and not our own cast related logic.

Perhaps move this to `AnyToCharVarchar` as static method, and make the length predicate a parameter?

Reply:
True, yes, I'll do that.